### PR TITLE
DateHistogram: Fix `extended_bounds` with `offset`

### LINKED ISSUE #23776: Date histogram: offset not working as expected
**Elasticsearch version**: 5.2.2

**Plugins installed**: none

**JVM version**: openjdk version "1.8.0_121"

**OS version**: docker container "elasticsearch:5.2.2"

**Description of the problem including expected versus actual behavior**:
Adding an offset to a date histogram with extended bounds does not behave like described in the documentation (https://www.elastic.co/guide/en/elasticsearch/reference/5.2/search-aggregations-bucket-datehistogram-aggregation.html#_offset).

Here's my request:
```
{
    'aggs': {
        'start_time': {
            'date_histogram': {
                'extended_bounds': {
                    'min': '2016-01-01T06:00:00Z',
                    'max': '2016-01-03T08:00:00Z'
                },
                'field': 'start_time',
                'interval': '1d',
                'min_doc_count': 0,
                'offset': '+6h'
            }
        }
    },
}
```

Here's the response I get (note the keys of the buckets):
```
{
    'aggregations': {
        'start_time': {
            'buckets': [
                {
                    'doc_count': 0,
                    'key': 1451606400000,
                    'key_as_string': '2016-01-01T00:00:00.000Z'
                },
                {
                    'doc_count': 3,
                    'key': 1451628000000,
                    'key_as_string': '2016-01-01T06:00:00.000Z'
                },
                {
                    'doc_count': 0,
                    'key': 1451692800000,
                    'key_as_string': '2016-01-02T00:00:00.000Z'
                },
                {
                    'doc_count': 0,
                    'key': 1451779200000,
                    'key_as_string': '2016-01-03T00:00:00.000Z'
                }
            ]
        }
    }
}
```

What I expected to get as a response (which I get using Elasticsearch 1.7):
```
{
    'aggregations': {
        'start_time': {
            'buckets': [
                {
                    'doc_count': 3,
                    'key': 1451628000000,
                    'key_as_string': '2016-01-01T06:00:00.000Z'
                },
                {
                    'doc_count': 0,
                    'key': 1451714400000,
                    'key_as_string': '2016-01-02T06:00:00.000Z'
                },
                {
                    'doc_count': 0,
                    'key': 1451800800000,
                    'key_as_string': '2016-01-03T06:00:00.000Z'
                }
            ]
        }
    }
}
```

I don't understand why I get so many buckets, and why do some of the buckets have starting date at midnight, even though I specified a `+6h` output.
If this is expected behaviour, how can I improve the request to get the result I got using Elasticsearch 1.7?

Thanks in advance for your help!

### PR DESCRIPTION
**Pull request by cbuescher**
This fixes a bug in 'date_histogram' when using 'extended_bounds'
together with some 'offset'. Offsets should be applied after rounding the
extended bounds and also be applied when adding empty buckets during the reduce
phase in InternalDateHistogram.

Closes #23776

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/ExtendedBounds.java [hunk 1]
```diff
@@ -166,8 +166,9 @@ ExtendedBounds parseAndValidate(String aggName, SearchContext context, DocValueF
         return new ExtendedBounds(min, max, minAsStr, maxAsStr);
     }
 
-    ExtendedBounds round(Rounding rounding) {
-        return new ExtendedBounds(min != null ? rounding.round(min) : null, max != null ? rounding.round(max) : null);
+    ExtendedBounds round(Rounding rounding, long offset) {
+        return new ExtendedBounds(min != null ? rounding.round(min) + offset : null,
+                max != null ? rounding.round(max) + offset : null);
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/ExtendedBounds.java]: should it be `rounding.round(X - offset) + offset` instead?

**Review by jpountz — COMMENTED** [review]
I left a question.

File: core/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/ExtendedBounds.java (hunk 1 above)
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/ExtendedBounds.java]: I don't think so. The min/max values should initially be the ones defined by the user in the query (e.g. "2016-01-01T04:00:00Z"). I we subtract the offset before rounding and e.g. do 1day intervals, we would get to 2015-12-31 in the example above, which I think is not the intention of the user. Does this make sense or am I missing something?

**Comment by cbuescher** [comment]
@jpountz I responded to your question, not sure if this makes sense. I could do the additional offset calculation in InternalDateHistogram#addEmptyBuckets() but this way its more centralized.

**Comment by jpountz** [comment]
I think part of the problem here is that the extended bounds are not well defined, eg. what is the expected behaviour if the provided bounds are not rounded values, should we round up or down? Maybe we should open a `discuss` issue about this.

**Comment by cbuescher** [comment]
@jpountz I opened #24064 to discuss this, but I think the current behaviour of always rounding down `extended_bounds` makes a lot of sense to me.

**Comment by cbuescher** [comment]
I rebased and changed this PR a bit so the offset isn't added in `ExtendedBounds#round()` but in `InternalDatehistogram#addEmptyBuckets`. I changed this because otherwise `ExtendedBounds#round()` wouldn't be an idempotent method, so mistakenly calling it several times with a set offset would have the potential to move the bounds further than intended. I think it is safer to keep the min/max bounds to be the rounded values and only fix the bug in the missing offset calculation in `addEmptyBuckets`
@jpountz could you have another look if this is better now?

**Review by jpountz — APPROVED** [review]

**Comment by kujon** [comment]
Could you please clarify which release was it included in in the end?

**Comment by cbuescher** [comment]
@kujon the attached labels indicate which versions this was merged to. So 5.5.0 means it is included in 5.5.x and all newer 5 version, and 6.0.0-alpha means its on all current 6.x releases. Hope this helps.

**Pull request merged**